alws/release_planner.py

```diff
@@ -170,6 +233,393 @@ async def get_pulp_packages(
+            selectinload(models.Release.product).selectinload(
+                models.Product.repositories),
+        ).with_for_update()
+        release_result = await self.db.execute(query)
+        release = release_result.scalars().first()
+        if not release:
+            raise DataNotFoundError(
+                f'Release with ID {release_id} not found')
+
+        if not can_perform(release, user, actions.ReleaseToProduct.name):
+            raise PermissionDenied('User does not have permissions '
+                                   'to update release')
+
+        build_tasks = getattr(payload, 'build_tasks', None)
+        if (payload.builds and payload.builds != release.build_ids) or (
+                build_tasks and build_tasks != release.build_task_ids):
+            release.build_ids = payload.builds
+            if build_tasks:
+                release.build_task_ids = payload.build_tasks
+            release.plan = await self.get_release_plan(
+                base_platform=release.platform,
+                build_ids=payload.builds,
+                build_tasks=payload.build_tasks,
+                product=release.product
+            )
+        elif payload.plan:
+            # check packages presence in prod repos
+            new_plan = await self.update_release_plan(
+                payload.plan, release)
+            release.plan = new_plan
+        self.db.add(release)
+        await self.db.commit()
+        await self.db.refresh(release)
+        logging.info('Successfully updated release %d', release_id)
+        return await self.__get_release(release.id)
+
+    async def commit_release(
+        self,
+        release_id: int,
+        user_id: int,
+    ) -> typing.Tuple[models.Release, str]:
+        logging.info('Commiting release %d', release_id)
+        users = await self.db.execute(select(models.User).where(
+            models.User.id == user_id).options(
+            selectinload(models.User.roles).selectinload(
+                models.UserRole.actions)
+        ))
+        user = users.scalars().first()
+
+        query = select(models.Release).where(
+            models.Release.id == release_id
+        ).options(
+            selectinload(models.Release.owner),
+            selectinload(models.Release.owner).selectinload(
+                models.User.roles).selectinload(models.UserRole.actions),
+            selectinload(models.Release.owner).selectinload(
+                models.User.oauth_accounts),
+            selectinload(models.Release.team).selectinload(
+                models.Team.roles).selectinload(models.UserRole.actions),
+            selectinload(models.Release.platform).selectinload(
+                models.Platform.repos.and_(
+                    models.Repository.production.is_(True)
+                ),
+            ),
+            selectinload(models.Release.product).selectinload(
+                models.Product.repositories
+            ),
+        ).with_for_update()
```

Comment:
can we move this query in function and use it here and in `update_release`?

Reply:
Moved